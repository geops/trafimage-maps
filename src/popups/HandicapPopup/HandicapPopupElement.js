import React from 'react';
import { useTranslation } from 'react-i18next';
import PropTypes from 'prop-types';

const propTypes = {
  properties: PropTypes.object.isRequired,
  propertyName: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
};

// 012 111 22 333
const simpleTelTester = /([0-9]{1,3}\s?[0-9]{1,3}\s?[0-9]{1,3}\s?[0-9]{1,3})/g;
// +41 (0) 11 222 33 44
const complexTelTester = /((([+]{1}[0-9]{1,3})|([+]?[(]{1}[0-9]{1,3}[)]{1})|([(]?[0-9]{4}[)]?))\s{0,4}[)]?[-\s\\.]?[(]?[0-9]{1,4}[)]?([^\r\n][-\s\\.]{0,1}[0-9]{1,3}){1,4})/g;

const emailTester = /[a-zA-Z0-9._+%-]+@[a-zA-Z0-9.-]+[.][a-zA-Z]{2,4}/gm;
const urlTester = /(www)?[a-z0-9]+([\\-\\.]{1}[a-z0-9]+)*\.[a-z]{2,5}(:[0-9]{1,5})?(\/.*)?/g;

const escapeRegExp = string => {
  return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
};

const searchMatches = (intialText, testers) => {
  let matched = [];
  testers.forEach(t => {
    const matches = intialText.match(t);
    if (matches) {
      matched.push(matches);
    }
  });
  // Flatten array of matches.
  if (matched.length) {
    matched = matched.flat();
  }

  if (matched) {
    // Remove duplicates from match array.
    return matched.sort().filter((item, pos, ary) => {
      return ary.indexOf(item) === pos;
    });
  }
  return [];
};

const mailTo = (email, idx) => (
  <a key={idx} href={`mailto:${email}`}>
    {email}
  </a>
);

const telTo = (telNumber, idx) => (
  <a key={idx} href={`tel:${telNumber}`}>
    {telNumber}
  </a>
);

const urlHref = (href, idx) => {
  return (
    <a
      key={idx}
      href={`http://${href}`}
      rel="noopener noreferrer"
      target="_blank"
    >
      {href}
    </a>
  );
};

const replaceLinks = (intialTextArray, matched, renderCallback) => {
  // Construct an RegExp to capture all matched emails.
  const regularExp = matched.reduce(
    (prevVal, currVal, currentIndex) =>
      currentIndex === 0 ? prevVal : `${prevVal}|(${escapeRegExp(currVal)})`,
    `(${escapeRegExp(matched[0])})`,
  );

  // Split array items and flatten it.
  const flatSplitArray = intialTextArray
    .map(item => {
      if (typeof item === 'string') {
        // Split string with RegExp and remove all empty strings or undefined.
        return item.split(new RegExp(regularExp, 'g')).filter(v => !!v);
      }
      return [item];
    })
    .flat();

  const newTextArray = flatSplitArray.map((text, idx) => {
    // eslint-disable-next-line react/no-array-index-key
    let substitutedElement = <span key={idx}>{text}</span>;
    matched.forEach(match => {
      if (text === match) {
        substitutedElement = renderCallback(match, idx);
      }
    });
    return substitutedElement;
  });
  return newTextArray;
};

const renderLinks = intialText => {
  const telMatches = searchMatches(intialText, [
    simpleTelTester,
    complexTelTester,
  ]);
  const emailMatches = searchMatches(intialText, [emailTester]);
  const urlMatches = searchMatches(intialText, [urlTester]);

  let replaced = [intialText];
  if (telMatches.length) {
    replaced = replaceLinks(replaced, telMatches, telTo);
  }
  if (emailMatches.length) {
    replaced = replaceLinks(replaced, emailMatches, mailTo);
  }
  if (urlMatches.length) {
    replaced = replaceLinks(replaced, urlMatches, urlHref);
  }
  return replaced;
};

function HandicapPopupElement({ properties, propertyName, label }) {
  const { t } = useTranslation();

  if (!properties[propertyName]) {
    return null;
  }

  const propLabel = label || propertyName;
  const values = properties[propertyName].split('\n');
  let content = null;

  if (values.length > 1) {
    content = (
      <>
        <div className="wkp-handicap-popup-field-title">{t(propLabel)}</div>
        <div className="wkp-handicap-popup-field-body">
          {values.map((v, idx) => (
            // eslint-disable-next-line react/no-array-index-key
            <div key={idx}>{renderLinks(v)}</div>
          ))}
        </div>
      </>
    );
  } else {
    content = (
      <>
        <div className="wkp-handicap-popup-field-title">{t(propLabel)}</div>
        <div className="wkp-handicap-popup-field-body">
          {renderLinks(values[0])}
        </div>
      </>
    );
  }

  if (!content) {
    return null;
  }

  return <div className="wkp-handicap-popup-element">{content}</div>;
}

HandicapPopupElement.propTypes = propTypes;

export default React.memo(HandicapPopupElement);
