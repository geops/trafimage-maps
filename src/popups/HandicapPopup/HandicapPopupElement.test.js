import React from 'react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import HandicapPopupElement from './HandicapPopupElement';
import 'react-app-polyfill/stable';

configure({ adapter: new Adapter() });

describe('HandicapPopupElement', () => {
  const props = {
    treffpunkt_de: '',
    zusaetzliche_informationen_de: `Test line here
      +33 666 66 66
      admin@email.it
      (text)
      LU-Tixi
      foo bar baz
      www.test.ch
      Dritte Dienstleistung
      Freiburg
      +41 (0) 41 210 00 60
      www.geops.ch`,
  };

  test('matches empty snapshot.', () => {
    const component = renderer.create(
      <HandicapPopupElement
        key="TreffPunkt"
        label="TreffPunkt"
        properties={props}
        propertyName="treffpunkt_de"
      />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("matches 'Zusätzliche Informationen' snapshot.", () => {
    const component = renderer.create(
      <HandicapPopupElement
        key="Zusätzliche Informationen"
        label="Zusätzliche Informationen"
        properties={props}
        propertyName="zusaetzliche_informationen_de"
      />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  [
    '061 777 66 66',
    '041 999 37 37',
    '+42 (0) 41 210 00 60',
    '056 632 12 21',
  ].forEach(number => {
    test(`should render telephone number '${number}' in href.`, () => {
      const wrapper = mount(
        <HandicapPopupElement
          key="test"
          label="test"
          properties={{
            test: number,
          }}
          propertyName="test"
        />,
      );
      const telHref = wrapper
        .find('a')
        .first()
        .getDOMNode();
      expect(telHref.getAttribute('href')).toBe(`tel:${number}`);
    });
  });

  ['test@test.fr', 'bla@lutixi.ch', 'foo@bluewin.ch'].forEach(email => {
    test(`should render email '${email}' in href.`, () => {
      const wrapper = mount(
        <HandicapPopupElement
          key="test"
          label="test"
          properties={{
            test: email,
          }}
          propertyName="test"
        />,
      );
      const telHref = wrapper
        .find('a')
        .first()
        .getDOMNode();
      expect(telHref.getAttribute('href')).toBe(`mailto:${email}`);
    });
  });

  test(`should handle several emails in the same input.`, () => {
    const wrapper = mount(
      <HandicapPopupElement
        key="test"
        label="test"
        properties={{
          test: 'test jean-jacques@hotmail.ch test another-address@email.zz',
        }}
        propertyName="test"
      />,
    );
    const firstTelHref = wrapper
      .find('a')
      .first()
      .getDOMNode();
    const secondTelHref = wrapper
      .find('a')
      .at(1)
      .getDOMNode();

    expect(firstTelHref.getAttribute('href')).toBe(
      'mailto:jean-jacques@hotmail.ch',
    );
    expect(secondTelHref.getAttribute('href')).toBe(
      'mailto:another-address@email.zz',
    );
  });

  ['www.geops.ch', 'www.test.ch/foo/bar/baz'].forEach(url => {
    test(`should render urls '${url}' in href.`, () => {
      const wrapper = mount(
        <HandicapPopupElement
          key="test"
          label="test"
          properties={{
            test: url,
          }}
          propertyName="test"
        />,
      );
      const telHref = wrapper
        .find('a')
        .first()
        .getDOMNode();
      expect(telHref.getAttribute('href')).toBe(`http://${url}`);
    });
  });
});
