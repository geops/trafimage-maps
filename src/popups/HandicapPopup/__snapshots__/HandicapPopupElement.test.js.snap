// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HandicapPopupElement matches 'Zusätzliche Informationen' snapshot. 1`] = `
<div
  className="wkp-handicap-popup-element"
>
  <div
    className="wkp-handicap-popup-field-title"
  >
    Zusätzliche Informationen
  </div>
  <div
    className="wkp-handicap-popup-field-body"
  >
    <div>
      Test line here
    </div>
    <div>
      <span>
              
      </span>
      <a
        href="tel:+33 666 66 66"
      >
        +33 666 66 66
      </a>
    </div>
    <div>
      <span>
        <span>
                
        </span>
      </span>
      <span>
        <a
          href="mailto:admin@email.it"
        >
          admin@email.it
        </a>
      </span>
    </div>
    <div>
            (text)
    </div>
    <div>
            LU-Tixi
    </div>
    <div>
            foo bar baz
    </div>
    <div>
      <span>
              
      </span>
      <a
        className="wkp-link"
        href="http://www.test.ch"
        rel="noopener noreferrer"
        target="_blank"
      >
        www.test.ch
        <svg>
          Link.svg
        </svg>
      </a>
    </div>
    <div>
            Dritte Dienstleistung
    </div>
    <div>
            Freiburg
    </div>
    <div>
      <span>
              
      </span>
      <a
        href="tel:+41 (0) 41 210 00 60"
      >
        +41 (0) 41 210 00 60
      </a>
    </div>
    <div>
      <span>
              
      </span>
      <a
        className="wkp-link"
        href="http://www.geops.ch"
        rel="noopener noreferrer"
        target="_blank"
      >
        www.geops.ch
        <svg>
          Link.svg
        </svg>
      </a>
    </div>
  </div>
</div>
`;

exports[`HandicapPopupElement matches empty snapshot. 1`] = `null`;
